main/src/ca/uwaterloo/flix/language/phase/Verifier.scala

```diff
@@ -235,6 +242,68 @@ object Verifier {
           check(expected = MonoType.Enum(sym.enumSym))(actual = t1, loc)
           tpe
 
+        case AtomicOp.ArrayLength =>
+          val List(arrt) = ts
+          arrt match {
+            case MonoType.Array(_) => check(expected = MonoType.Int32)(actual = tpe, loc)
+            case _ => throw MismatchedShape(tpe, s"An Array", loc)
+          }
+
+        case AtomicOp.ArrayNew =>
+          val List(elmt, lent) = ts
+          check(expected = MonoType.Int32)(actual = lent, loc)
+          check(expected = MonoType.Array(elmt))(actual = tpe, loc)
+
+        case AtomicOp.ArrayLit =>
+          tpe match {
+            case MonoType.Array(elmt) =>
+              ts.foreach(t => check(expected = elmt)(actual = t, loc))
+            case _ => throw MismatchedShape(tpe, s"An Array", loc)
+          }
+          tpe
+
+        case AtomicOp.ArrayLoad =>
+          val List(arrt, idxt) = ts
+          check(expected = MonoType.Int32)(actual = idxt, loc)
+          arrt match {
+            case MonoType.Array(elmt) => check(expected = elmt)(actual = tpe, loc)
+            case _ => throw MismatchedShape(tpe, s"An Array", loc)
+          }
+          tpe
+
+        case AtomicOp.ArrayStore =>
+          val List(arrt, idxt, expt) = ts
+          check(expected = MonoType.Int32)(actual = idxt, loc)
+          arrt match {
+            case MonoType.Array(elmt) => check(expected = elmt)(actual = expt, loc)
+            case _ => throw MismatchedShape(tpe, s"An Array", loc)
+          }
+          check(expected = MonoType.Unit)(actual = tpe, loc)
+
+        case AtomicOp.Ref =>
+          val List(elm) = ts
+          check(expected = MonoType.Ref(elm))(actual = tpe, loc)
+
+        case AtomicOp.Deref =>
+          val List(ref) = ts
+          ref match {
+            case MonoType.Ref(elm) => check(expected = elm)(actual = tpe, loc)
+            case _ => throw MismatchedShape(tpe, s"Ref", loc)
+          }
+
+        case AtomicOp.Lazy =>
+          val List(elm) = ts
+          check(expected = MonoType.Lazy(elm))(actual = tpe, loc)
+
+        case AtomicOp.Force =>
+          val List(laz) = ts
+          laz match {
+            case MonoType.Lazy(elm) => check(expected = elm)(actual = tpe, loc)
+            case _ => throw MismatchedShape(tpe, s"Lazy", loc)
+          }
+
+        case AtomicOp.Tuple => check(expected = MonoType.Tuple(ts))(actual = tpe, loc)
```

Comment:
`AtomicOp.Index(idx: Int)` is the inverse of tuple (extract the idx'th value from the tuple), so I think you should add that. It's a bit different from the rest because you have to look into the tuple type of the exp but give it a shot :)